From c7747b6f1cb8dffeb887a2b09755c69ef4484a1f Mon Sep 17 00:00:00 2001
From: Anusha Ramineni <anusha.ramineni@india.nec.com>
Date: Fri, 31 May 2019 15:47:12 +0530
Subject: [PATCH] Fix GetDiskFormat to retrieve o/p in case of Alpine

blkid o/p in Alpine is single line which is not considered
while retrieving disk format. This commit is to update the same.
---
 pkg/util/mount/mount_linux.go      | 11 ++--
 pkg/util/mount/mount_linux_test.go | 87 ++++++++++++++++++++++++++++++
 2 files changed, 95 insertions(+), 3 deletions(-)

diff --git a/vendor/k8s.io/kubernetes/pkg/util/mount/mount_linux.go b/pkg/util/mount/mount_linux.go
index 0f75d2184d55b..6e93b913897cf 100644
--- a/vendor/k8s.io/kubernetes/pkg/util/mount/mount_linux.go
+++ b/vendor/k8s.io/kubernetes/pkg/util/mount/mount_linux.go
@@ -514,7 +514,7 @@ func (mounter *SafeFormatAndMount) GetDiskFormat(disk string) (string, error) {
 	args := []string{"-p", "-s", "TYPE", "-s", "PTTYPE", "-o", "export", disk}
 	klog.V(4).Infof("Attempting to determine if disk %q is formatted using blkid with args: (%v)", disk, args)
 	dataOut, err := mounter.Exec.Run("blkid", args...)
-	output := string(dataOut)
+	output := strings.TrimSpace(string(dataOut))
 	klog.V(4).Infof("Output: %q, err: %v", output, err)
 
 	if err != nil {
@@ -534,6 +534,10 @@ func (mounter *SafeFormatAndMount) GetDiskFormat(disk string) (string, error) {
 	var fstype, pttype string
 
 	lines := strings.Split(output, "\n")
+	if len(lines) == 1 {
+		// In case of Alpine, o/p will be one line. So, try to split by space
+		lines = strings.Split(output, " ")
+	}
 	for _, l := range lines {
 		if len(l) <= 0 {
 			// Ignore empty line.
@@ -541,12 +545,13 @@ func (mounter *SafeFormatAndMount) GetDiskFormat(disk string) (string, error) {
 		}
 		cs := strings.Split(l, "=")
 		if len(cs) != 2 {
-			return "", fmt.Errorf("blkid returns invalid output: %s", output)
+			// If it's not key=value pair, ignore
+			continue
 		}
 		// TYPE is filesystem type, and PTTYPE is partition table type, according
 		// to https://www.kernel.org/pub/linux/utils/util-linux/v2.21/libblkid-docs/.
 		if cs[0] == "TYPE" {
-			fstype = cs[1]
+			fstype = strings.ReplaceAll(cs[1], "\"", "")
 		} else if cs[0] == "PTTYPE" {
 			pttype = cs[1]
 		}
